[tidy] add new check bugprone-return-const-ref-from-parameter (#89497)
llvm/llvm-project: clang-tools-extra/docs/clang-tidy/checks/bugprone/return-const-ref-from-parameter.rst

.. title:: clang-tidy - bugprone-return-const-ref-from-parameter

bugprone-return-const-ref-from-parameter
========================================

Detects the function which returns the const reference from parameter which
causes potential use after free if the caller uses xvalue as argument.

> In c++, const reference parameter can accept xvalue which will be destructed

Reviewer: ```suggestion
In C++, const reference parameter can accept xvalue which will be destructed
```

> In c++, const reference parameter can accept xvalue which will be destructed
> after the call. When the function returns this parameter also as const reference,
> then the returned reference can be used after the object it refers to has been
> destroyed.

Reviewer 2: ```suggestion
In C++, constant reference parameters can accept xvalues which will be destructed
after the call. When the function returns such a parameter also as constant reference,
then the returned reference can be used after the object it refers to has been
destroyed.
```

.. title:: clang-tidy - bugprone-return-const-ref-from-parameter

bugprone-return-const-ref-from-parameter
========================================

> Detects return statements that return constant reference parameter as constant

Reviewer 2: ```suggestion
Detects return statements that return a constant reference parameter as constant
```

> Detects return statements that return a constant reference parameter as constant
> reference. This may cause use-after-free errors if the caller uses xvalue as
> arguments.

Reviewer 2: ```suggestion
Detects return statements that return a constant reference parameter as constant
reference. This may cause use-after-free errors if the caller uses xvalues as
arguments.
```
